File: src/main/java/net/bbelovic/adventofcode/day1/Day1PuzzlePart2.java
package net.bbelovic.adventofcode.day1;

import net.bbelovic.adventofcode.Puzzle;

public class Day1PuzzlePart2 implements Puzzle<String, Long> {

    public Long solve(String input) {
        long position = 0L;
        long result = 0L;
        for (char c: input.toCharArray()) {
            if (c == ')') {
                result = result - 1;
            } else if (c == '(') {
                result = result + 1;
            }
            ++position;
            if (result == -1) {
                break;
            }
        }
        return position;
    }
}

File: aoc-2018/src/main/kotlin/net/bbelovic/adventofcode/year2018/day1/Day1PuzzlePart2.kt
package net.bbelovic.adventofcode.year2018.day1

import net.bbelovic.adventofcode.Puzzle

class Day1PuzzlePart2 : Puzzle<String, Long> {
    override fun solve(input: String): Long {
        var notFound = true
        var startingFrequency = 0L
        val knownFrequencies = mutableSetOf(startingFrequency)
        val frequencyList = input.splitToSequence(", ").map { it -> it.toLong() }.toList()
        var result = Long.MIN_VALUE
        while (notFound) {
            for (frequency in frequencyList) {
                val candidate = startingFrequency + frequency
                if (candidate in knownFrequencies) {
                    result = candidate
                    notFound = false
                    break
                } else {
                    knownFrequencies.add(candidate)
                }
                startingFrequency = candidate
            }
        }
        return result
    }
}